ports/esp32s2/common-hal/countio/Counter.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 microDev
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "common-hal/countio/Counter.h"
+#include "common-hal/microcontroller/Pin.h"
+
+void common_hal_countio_counter_construct(countio_counter_obj_t* self,
+    const mcu_pin_obj_t* pin) {
+    claim_pin(pin);
+
+    // Prepare configuration for the PCNT unit
+    pcnt_config_t pcnt_config = {
+        // Set PCNT input signal and control GPIOs
+        .pulse_gpio_num = pin->number,
+        .ctrl_gpio_num = PCNT_PIN_NOT_USED,
+        .channel = PCNT_CHANNEL_0,
+        // What to do on the positive / negative edge of pulse input?
+        .pos_mode = PCNT_COUNT_INC,   // Count up on the positive edge
+        .neg_mode = PCNT_COUNT_DIS,   // Keep the counter value on the negative edge
+    };
+    // Initialize PCNT unit
+    pcnt_handler_init(&pcnt_config);
```

Comment:
How about returning the unit number? I think that'd be clearer: `self->unit = pcnt_handler_init(&pcnt_config)`

Reply:
I don't have a solid reason for not using return but there are some scenarios where the current implementation would be more clear.

Reply:
Why is the current implementation clearer? I'd prefer a return and you can use a value < 0 to indicate an error instead of raising an exception from inside the function. Generally, `peripherals` code shouldn't include any MicroPython structures including exceptions so that the code can be used outside the VM.

Reply:
I was looking at this with the perspective of limiting the visibility setting unit to `pcnt.c` only... using return now.